docs/design/proposals/rbd-mirror.md

```diff
@@ -0,0 +1,102 @@
+
+> Sample PV spec which will be used by rbdplugin controller to regenerate OMAP
+> data
+
+```yaml
+csi:
+  controllerExpandSecretRef:
+    name: rook-csi-rbd-provisioner
+    namespace: rook-ceph
+  driver: rook-ceph.rbd.csi.ceph.com
+  fsType: ext4
+  nodeStageSecretRef:
+    name: rook-csi-rbd-node
+    namespace: rook-ceph
+  volumeAttributes:
+    clusterID: rook-ceph
+    imageFeatures: layering
+    imageFormat: "2"
+    imageName: csi-vol-0c23de1c-18fb-11eb-a903-0242ac110005
+    journalPool: replicapool
+    pool: replicapool
+    radosNamespace: ""
+  volumeHandle: 0001-0009-rook-ceph-0000000000000002-0c23de1c-18fb-11eb-a903-0242ac110005
+```
+
+> **VolumeHandle** is the unique volume name returned by the CSI volume plugin’s
+CreateVolume to refer to the volume on all subsequent calls.
+
+Once the static PVC is created on the secondary cluster, the Kubernetes User
+can try delete the PVC,expand the PVC or mount the PVC. Incase of mounting
+(NodeStageVolume) we will get the volume context in RPC call but not in the
+Delete/Expand Request. In Delete/Expand RPC request only the VolumeHandle
+(`clusterID-poolID-volumeuniqueID`) will be sent where it contains the encoded
+information related to ClusterID and PoolID. The VolumeHandle is not useful in
+secondary cluster as the PoolID and ClusterID always may not be the same.
+
+> In this design document we will talk about new controller(rbdplugin
+> controller) not replication controller, in next releases we will design the
+> replication controller to perform mirroring operations. The rbdplugin
+> controller will runs as a sidecar in RBD provisisoner pod.
+
+To solve this problem, We will have a new controller(rbdplugin controller)
+running as part of provisioner pod which watchers for the PV objects. When a PV
+is created it will extract the required information from the PV spec and it
+will regenerate the OMAP data and also it will generate a new VolumeHandle
+(`newclusterID-newpoolID-volumeuniqueID`) and it creates an OMAP object for
+mapping between old VolumeHandle and new VolumeHandle. whenever Ceph-CSI gets a
+RPC request with older VolumeHandle it will check is there any new VolumeHandle
+exists for the old VolumeHandle. If yes, it uses the new VolumeHandle for
+internal operations (to get pool name, Ceph monitor details from the ClusterID
+etc).
+
+Currently, We are making use of watchers in node stage request to make sure
+ReadWriteOnce (RWO) PVC is mounted on a single node at a given point in time.
+We need to change the watchers logic in the node stage request as when we
+enable the RBD mirroring on an image, a watcher will be added on a RBD image by
+the rbd mirroring daemon.
+
+To solve the ClusterID problem, If the ClusterID is different on the second
+cluster the admin as to create a new ConfigMap for the mapped ClusterID's.
+whenever Ceph-CSI gets a request it will check the ClusterID mapping exists and
+uses mapped ClusterID to get the information like Ceph monitors etc.
```

Comment:
```suggestion
Whenever Ceph-CSI gets a request, it will check if the ClusterID mapping exists and
uses the mapped ClusterID to get the information like Ceph monitors etc.
```